### DISCUSSION TITLE
Is it possible to write files from within a plugin?

### DISCUSSION THREAD

**Discussion by ole-treichel** [swc-project/swc#4997] [Q&A] [answered]
I want to use a plugin to extract things from a JS file (e.g. ˋpropTypesˋ) and write them to a seperate file on disk.

**Comment by kwonoj** [comment, marked as answer]
There is a limited filesystem access for `/cwd` to current working directory, if plugin is compiled as wasi target.

**Reply by ole-treichel** [reply]
Awesome 👍  could you point me to any example / reference code? I searched the repo but did not find anything relevant yet.

**Reply by kwonoj** [reply]
There isn't so far I know, except PR enables `/cwd` alias itself if you'd like to see how it is exposed. Using fs access is not much different anyway other than it only allows known explicit path alias.

**Reply by ole-treichel** [reply]
Thanks. I still can't make it work, sorry if I misunderstood. I tried the following in my plugin:

```rust
let data = "Some data";
let out_path = env::current_dir().expect("Failed to get cwd").join("test.txt");

println!("out_path: {}", out_path.display());

fs::write(out_path, data).expect("Unable to write file");
```

This does not work, permission is denied:

```thread '<unnamed>' panicked at 'Unable to write file: Os { code: 2, kind: PermissionDenied, message: "Permission denied" }'```

I can also see that `cwd` is just `/`. I tried hardcoding the path to the project directory, but no luck either. It cannot find the directory then (which I guess makes sense if the `cwd` is "mounted" inside `wasi` as `/`.

I am using (almost) the latest swc crates / npm-packages:

```
swc_plugin = "=0.57.0"
serde = "1.0.117"
```

```
"devDependencies": {
    "@swc/cli": "0.1.57",
    "@swc/core": "1.2.204"
}
```

**Reply by kwonoj** [reply]
You should use mounted dir, any attempt to calculate actual physical path including env::current_dir is expected to fail. Since plugin runs in the sandboxed context, you only get access to current dir's mounted alias.

**Reply by ole-treichel** [reply]
Could you give an example with `fs`? Sorry, I don't really understand what you mean by "mounted" directory, is this a special path?

**Reply by ole-treichel** [reply]
Nvm got it: the mounted directory is `/cwd`, so this works:

```rust
let data = "Some data";

println!("out_path: {}", out_path.display());

fs::write("/cwd/test.txt", data).expect("Unable to write file");
```

Appreciate your help!

**Comment by romanusname** [comment]
I can't add text to a file. `fs::write("/cwd/test.txt", data).expect("Unable to write file");` completely overwrites it. And when I try to read I get an error. Can you add new text to a file?

**Reply by kwonoj** [reply]
https://doc.rust-lang.org/std/fs/fn.write.html

> This function will create a file if it does not exist, and will entirely replace its contents if it does.

**Reply by romanusname** [reply]
> https://doc.rust-lang.org/std/fs/fn.write.html
> 
> > This function will create a file if it does not exist, and will entirely replace its contents if it does.

"Write a slice as the entire contents of a file.". The file says "Hello", you need to add "World". CWS does not read files. I created a separate question. Look here please. i really need help https://github.com/swc-project/swc/discussions/7359

**Reply by kwonoj** [reply]
Now I understood q, it's not about fs more about the approach: if components are separated into different files, no, there's no `global` lifetime you can expect to store-at-read once some states.

**Reply by romanusname** [reply]
> Now I understood q, it's not about fs more about the approach: if components are separated into different files, no, there's no `global` lifetime you can expect to store-at-read once some states.


Can't read the file? Or am I doing something wrong? `let r = fs::read("test.txt").expect("test");` test without error but next js `thread '<unnamed>' panicked at 'failed to invoke plugin...`
